File: app/android/domain/PriorityLogic.kt
package com.primafocus.domain

import kotlin.math.ln
import kotlin.math.max

object PriorityLogic {

    fun calculatePriorityScore(
        categoryWeight: Double,
        hasDate: Boolean,
        timeUrgency: Double,
        subtasksCount: Int,
        estimatedMinutes: Int,
        ageDays: Double,
        manualBoost: Double = 5.0
    ): Double {
        val hasDateInt = if (hasDate) 1 else 0
        
        var score = (10 * categoryWeight) +
                (6 * hasDateInt) +
                (8 * timeUrgency) -
                (2 * ln(1.0 + subtasksCount)) -
                (0.02 * estimatedMinutes) -
                (0.5 * ageDays) +
                manualBoost

        if (categoryWeight >= 4.0) {
            score = max(score, 70.0)
        }

        return score
    }

    fun evaluateIsProject(estimatedMinutes: Int, subtasksCount: Int): Boolean {
        return estimatedMinutes > 180 || subtasksCount > 10
    }
}

File: app/web/src/db/priorityLogic.ts
import type { Task } from './models';

export function calculatePriorityScore(task: Partial<Task>, ageDays: number = 0): number {
  const categoryWeight = task.categoryWeight ?? 1.0;
  const hasDate = task.date ? 1 : 0;
  const timeUrgency = task.timeUrgency ?? 0.0;
  const subtasksCount = task.subtasksCount ?? 0;
  const estimatedMinutes = task.estimatedMinutes ?? 0;
  const manualBoost = task.manualBoost ?? 5.0;

  let score = 
    10 * categoryWeight + 
    6 * hasDate + 
    8 * timeUrgency - 
    2 * Math.log(1 + subtasksCount) - 
    0.02 * estimatedMinutes - 
    0.5 * ageDays + 
    manualBoost;

  if (categoryWeight >= 4.0) {
    score = Math.max(score, 70);
  }

  return score;
}

export function evaluateIsProject(task: Partial<Task>): boolean {
  const estimatedMinutes = task.estimatedMinutes ?? 0;
  const subtasksCount = task.subtasksCount ?? 0;
  return estimatedMinutes > 180 || subtasksCount > 10;
}
